### ISSUE TITLE
How to narrow down default "read-only" role? 

### ISSUE THREAD

**Issue by dan-m8t** [argoproj/argo-cd#9900]
Hi there,

I got a problem in ArgoCD regarding RBAC.
I am using OKD 4.8 with the ArgoCD Operator and ArgoCD in version 2.3.4.

We operate a multi-tenant cluster and want to narrow down the permissions as far as possible.

By default every authenticated user has read-only rights and can see anything inside ArgoCD. This is a behaviour we do not want, as we already seperate projects by namespaces with corresponding access groups in OKD this should be the same on ArgoCD.

At first we had the following approach (Operator Config):

```rbac:
     defaultPolicy: ""
     policy: |
       g, cluster-admins, role:admin
     scopes: '[groups]'
```

This config removes the read-only policy and grants cluster-admins full administrative permissions in ArgoCD.

We are defining permissions via AppProjects, every AppProject gets an "admin" and a "view" group.

After I removed the defaultPolicy RBAC project admins (defined by AppProjects) can see their projects as before (those project admins are NOT cluster-admins!), the members of the viewer group are not able to see anything anymore.

I tried to work with the AppProject rights but it seems that an ArgoCD user needs a "get" cluster-wide I guess?

Here is an example of the permissions we use inside a AppProject:

```apiVersion: argoproj.io/v1alpha1
 kind: AppProject
 metadata:
   name: prj-xxx-xx
   namespace: argocd-devops
 spec:
   clusterResourceWhitelist:
   - group: '*'
     kind: '*'
   description: some project
   destinations:
   - namespace: prj-xxx-xx-dev
     server: https://kubernetes.default.svc
   namespaceResourceWhitelist:
   - group: '*'
     kind: '*'
   roles:
   - groups:
     - prj-xxx-xx-dev-admins
     name: prj-xxx-xx-dev-admins
     policies:
     - p, proj:prj-xxx-xx:prj-xxx-xx-dev-admins, applications, *, prj-xxx-xx/dev*, allow
     - p, proj:prj-xxx-xx:prj-xxx-xx-dev-admins, applications, get, prj-xxx-xx/ops*, allow
   - groups:
     - prj-xxx-xx-dev-viewer
     name: prj-xxx-xx-dev-viewer
     policies:
     - p, proj:prj-xxx-xx:prj-xxx-xx-dev-viewer, applications, get, prj-xxx-xx/*, allow
   sourceRepos:
   - https://repo1.bar/bla.git
   - https://repo2.foo/bla2.git
  ```

The group prj-xxx-xx-dev-admins has all the rights inside the AppProject for applications starting with "dev" and "get" rights for the applications starting with "ops*" (those are default settings we set for an OKD project)

The viewer group prj-xxx-xx-dev-viewer has permissions for the whole project to "get" ressources <-- I think this is unnecessary because the default read-only role would grant those permissions anyways, am I right?

What I tried so far to copy/paste the default "read-only" global role permissions into the AppProject, see here for reference: https://github.com/argoproj/argo-cd/blob/master/assets/builtin-policy.csv

This didn't work either as it seems there is missing a permission "cluster-wide" I guess? What I don't understand is that project admins are able to see their projects (and only the projects they have permissions to!) after I removed the defaultPolicy in the operator.

My next guess was to create a narrowed down global cluster read-only role, but I was not able to figure out if this is possible somehow? Which permissions exactly is needed to view projects (see default read-only).

How can I achieve this? I hope it is clear what I want to do :) 

Thanks

Dan

**Comment by crenshaw-dev** [comment]
> Good to know ;) Do you have a "defaultPolicy" defined?

I'm using vanilla Argo CD, so I set `policy.default: ""` in argocd-rbac-cm.  I'm guessing that's what OKD's `defaultPolicy` does?

I even restarted the API server just to be sure it took effect.

I must admit, I'm a bit stumped. I think I'd have to use a debugger and step through the auth code to figure out why the Applications aren't showing up.

**Comment by dan-m8t** [comment]
> I'm using vanilla Argo CD, so I set policy.default: "" in argocd-rbac-cm. I'm guessing that's what OKD's defaultPolicy does?

The ArgoCD operator sets "read-only" as the defaultPolicy

> I must admit, I'm a bit stumped. I think I'd have to use a debugger and step through the auth code to figure out why the Applications aren't showing up.

Is there something I can provide to debug the problem?

**Comment by dan-m8t** [comment]
I have to admit that me and my colleagues screwed up :) 

I didn't post here the output of `oc get groups | grep prj-xxx-xx` - that would have solve this way earlier.


Turns out we named the viewer groups like this: prj-xxx-xx-stage-viewer**s** (mind the s :)).

In our AppProjects we always set a permission for "get" for the applications bind to the group membership as mentioned above. But we forgot to append the **s** all the time. And because until we change the defaultPolicy to "none" nobody took note because the viewers were able to view their (and all the other) projects. That's a hard one, but it works now lol.

Thanks for the help ;)
